src/contracts/core/Slasher.sol

```diff
@@ -5,233 +5,475 @@ import "./SlasherStorage.sol";
-     * operator, strategy, epoch, and operator set assuming no further
-     * modifications to requested slashing rate by operatorSet
-     */
-    function getPendingSlashingRate(
+    /// @notice push or update the latest TotalAndNonslashable update for the given operator, strategy
+    /// depending on the current effectTimestamp
+    function _updateTotalAndNonslashableHistory(
         address operator,
         IStrategy strategy,
-        IOperatorSetManager.OperatorSet calldata operatorSet,
-        uint32 epoch
-    ) external view returns (uint32) {
-        uint32 requestedSlashingRate = getRequestedSlashingRate(operator, strategy, operatorSet, epoch);
-        if (requestedSlashingRate >= SlashingAccountingUtils.BIPS_FACTOR_SQUARED) {
-            requestedSlashingRate = uint32(SlashingAccountingUtils.BIPS_FACTOR_SQUARED);
+        TotalAndNonslashableUpdate storage totalAndNonslashableUpdate,
+        uint64 newTotalMagnitude,
+        uint64 newNonslashableMagnitude,
+        uint32 effectTimestamp
+    ) internal {
+        if (totalAndNonslashableUpdate.timestamp == effectTimestamp) {
+            totalAndNonslashableUpdate.totalMagnitude = newTotalMagnitude;
+            totalAndNonslashableUpdate.nonSlashableMagnitude = newNonslashableMagnitude;
+        } else {
+            _totalMagnitudeUpdates[operator][strategy].push(TotalAndNonslashableUpdate({
+                timestamp: effectTimestamp,
+                totalMagnitude: newTotalMagnitude,
+                nonSlashableMagnitude: newNonslashableMagnitude
+            }));
         }
-        return requestedSlashingRate;
     }
 
-    /**
-     * @notice fetches the parts per hundred million that will be slashed for
-     * the given operator, strategy, and epoch, across all operator set assuming
-     * no more modifications to requested slashing rate for the operator.
-     *
-     * @param operator the operator to get the pending slashing rate for
-     * @param strategy the strategy to get the pending slashing rate for
-     * @param epoch the epoch to get the pending slashing rate for
-     *
-     * @return the parts per hundred million that will be slashed for the
-     * given operator, strategy, and epoch, across all operator set assuming
-     * no more modifications to requested slashing rate for the operator.
-     */
-    function getTotalPendingSlashingRate(
+    /// @notice Rate limit number of allocations/deallocation updates to 3
+    /// for a given (operator, strategy, and operatorSet)
+    function _checkPendingUpdates(
         address operator,
         IStrategy strategy,
-        uint32 epoch
-    ) external view returns (uint32) {
-        uint64 totalSlashingRate = slashingRequests[operator][strategy][epoch].slashingRate;
-        if (totalSlashingRate >= SlashingAccountingUtils.BIPS_FACTOR_SQUARED) {
-            totalSlashingRate = SlashingAccountingUtils.BIPS_FACTOR_SQUARED;
+        OperatorSet calldata operatorSet
+    ) public view {
```

Comment:
public?

Reply:
if we want to have this public, pls take of the `_`

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/623/commits/f7c9a75affd75db460bf53ff9dc6d5b8ff9d32db